jwst/associations/lib/rules_level3.py

```diff
@@ -604,7 +604,7 @@ def __init__(self, *args, **kwargs):
         ])
```

Comment:
Looks to me like the inclusion of `Constraint_MSA()` a few lines above means this is also going to generate spec3 ASN's for NRC_MSASPEC observations, hence the rule name `NRCWFSS` seems too specific. Should we pull the NIRSpec stuff into a separate (new) rule?

Reply:
Hmmmm, maybe jumped the gun on the renaming. I was just looking at the primary exp_type check.

So, this rule originally handled all "source-focused" associations. Then niriss wfss got removed and I mistakenly thought that was all there was to it.

Two ways to go about this: Either leave the original rule as is, except for commentary about "except NIRISS WFSS". Or split out and make a rule for each of the modes. Original philosophy was to make the rules as "science-focused" as possible and not "instrument-mode-focused", so the my preference is the former option.

Reply:
The existing `WFSSNIS` rule has constraints on exp_type ('nis_wfss'), filter ('gr150r', 'gr150c'), and pupil. Meanwhile the `SpectralSource` rule being addressed here only seems to use exp_type ('nrc_wfss') to select NIRCam WFSS exposures. Any ideas as to why the one for NRC is so simple and generic, while the one for NIS is more specific? Just wondering how hard it might be to create a generic `WFSS` rule that could apply to both NRC and NIS and then have a separate one for NRS MSASPEC.

Alternatively, is there a way to combine all 3 into a single rule that covers all multi-object modes (NRS MOS, NRC WFSS, and NIS WFSS)?

Reply:
About the last point: One could combine, but that would create an extremely complicated rule, the main differential being the grism/catalog aspect. Not sure that would ever be de-buggable.

About why NIRISS Grism is different: Would have to explore more, but the split occurred at PR #1983.

Reply:
It looks to me like we had a single, generic `Asn_WFSS` rule, which was not actually completely generic, because it only included exp_type `nis_wfss`, up until https://github.com/spacetelescope/jwst/pull/2001 was introduced, which renamed it to the specific `Asn_WFSSNIS` rule. I'm guessing we had the unique rule for nis_wfss due to issues with using the filter and pupil values to construct the appropriate opt_elem pieces of the product name(s) and perhaps also for use in determining the correct corresponding catalog and direct image names?

But given that NIRCam also uses filter and pupil, why don't we need those special hooks for nrc_wfss? Perhaps it's because the NIRISS grisms happen to be physically installed in the filter wheel, while for NIRCam they're in the pupil wheel?

Reply:
Just for history/context: The original idea was to have just two spectrographic rules:

- Single Object/target-based
- Mulit-object/source-based

Then reality started hitting. Fix Slit/Slitless, whichever mode is the source-based one, had to be separated out. Then WFSS was actually finalized, originally based on NIRISS, which was called Asn_WFSS. The presumption was that NIRCam would also fall under there, but either didn't need to or possibly it should but just hasn't been migrated. The final "source" mode is IFU, but that either never made it in or was determined to be to different to be grouped.

So, with 20-20 hindsight, the ideal groupings would be, again trying to be science-based as opposed to instrument:

- Single Object/Target
- IFU
- MSA
- Fixed slit/slitless (which ever is the source one)
- Grism/catalog

and the suggestion on naming would be `Asn_Lv3[mode][instrument]` where `instrument` is the optional part, reverse of what was suggested above. 

Methinks this issue should just be closed out and a new issue (again methinks there may be an old github issue already) created to deal with just the overall rename/refactor. There is an issue about removing the redundant `Asn_` prefix; tech debt from a previous version.

Reply:
I agree with closing this PR and related ticket/issue and filing a new ticket aimed at a bit more ambitious refactoring/reorganizing of the various spectroscopic rule classes. This PR is not needed to fix an existing problem, it was only for the purpose of doing some clean-up.